### PR TITLE
Adding cascading deletion support to federated namespaces

### PR DESCRIPTION
**Pull request by nikhiljindal**
Ref https://github.com/kubernetes/kubernetes/issues/33612

With this change, whenever a federated namespace is deleted with `DeleteOptions.OrphanDependents = false`, then federation namespace controller first deletes the corresponding namespaces from all underlying clusters before deleting the federated namespace.

cc @kubernetes/sig-cluster-federation @caesarxuchao

``` release-note
Adding support for DeleteOptions.OrphanDependents for federated namespaces. Setting it to false while deleting a federated namespace also deletes the corresponding namespace from all registered clusters.
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34648)

### CODE REVIEW

**Comment by nikhiljindal** [comment]
I am working on updating our e2e test

**Comment by nikhiljindal** [comment]
Updated the PR to include e2e changes. Verified that test/e2e/federated-namespace passes

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (caesarxuchao) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: I thought ProcessDeletion should take a `fed_v1.Namespace`, why is it taking a regular namespace?

**Review by caesarxuchao — COMMENTED** [review]
Sorry, I'm not familiar with the federation code base. I asked two question, they might be silly. I'll take another pass tomorrow.

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Fixed, thx

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: There is no special `fed_v1.Namespace`. We use the same v1.Namespace object in federation

**Comment by nikhiljindal** [comment]
Updated the code as per comments. Ready for another look

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: No. federated namespace will not be deleted unless the namespaces have been deleted in all underlying clusters.

Comment (caesarxuchao) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Where is it blocked? It seems the [UpdateOnError](https://github.com/kubernetes/kubernetes/pull/34648/files#diff-2c30b9c3878f0bf2ea184f5d3b9c30c3R99) is not blocking.

Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Yes it was blocked until the Delete request succeeded for all clusters but it wasnt waiting for all resources to go away. Updated the code to wait.

**Comment by nikhiljindal** [comment]
Updated the controller to add a DeleteFederatedDependents finalizer before creating any dependent resources in sub clusters. This now ensures that the federation resource is not deleted unless all dependents have been removed.
PTAL.

**Comment by nikhiljindal** [comment]
Updated the code as discussed. I am now creating the finalizer in ObjectMeta.Finalizers instead of putting it in NamespaceSpec.Finalizers.
Verified that federated-namespace e2e test passes :)

**Comment by nikhiljindal** [comment]
Retrying unit test
@k8s-bot unit test this

**Comment by nikhiljindal** [comment]
Retrying unit test
@k8s-bot unit test this

File: federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go
Comment (caesarxuchao) [inline federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go]: Where is the "wait" is implemented?

File: pkg/api/v1/types.go
Comment (caesarxuchao) [inline pkg/api/v1/types.go]: Thanks!

File: test/e2e/federated-namespace.go
Comment (caesarxuchao) [inline test/e2e/federated-namespace.go]: What is Federation12?

File: test/e2e/framework/util.go
Comment (caesarxuchao) [inline test/e2e/framework/util.go]: Do you want to check in this code?

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (caesarxuchao) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Why not just call Update? Calling `Finalize` looks like it's delete the namespace.spec.finalizer

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (caesarxuchao) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: I think you should check if all resources are actually deleted to make the deletion synchronous. Cluster namespace deletion will wait for all resources are deleted. Perhaps this will be easier to implement after you apply the deleteHelper pattern to other resources?

File: federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go
Comment (caesarxuchao) [inline federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go]: Add a comment? Also mention the OrphanFinalizer will take precedence.

**Review by caesarxuchao — CHANGES_REQUESTED** [review]

File: test/e2e/framework/util.go
Comment (nikhiljindal) [inline test/e2e/framework/util.go]: Yes helps in debugging.

File: federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go]: Wait is implemented by not returning success here.
Since `len(operations) > 0` we return here without removing the finalizer. Next time this function is called, we will again first compute if we want to do any operations in underlying clusters. If at any time, len(operation) = 0, then we proceed. Else we wait till len(operations) == 0.
Caller is expected to keep calling HandleObjectInUnderlyingClusters until it returns success (err == nil).

File: federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/util/deletionhelper/deletion_helper.go]: done, thx

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: > I think you should check if all resources are actually deleted to make the deletion synchronous. 

Yes this is not required right now since all deletions are immediate (there is no cascading deletion). Will need to update this before enabling cascading deletion for any of these resources.
The plan is to refactor kubernetes namespace controller code to make the namespace deletion code reusable and then use it here.
Leaving as is in this PR.

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Yes nil should work as well. Will try :)

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Yes, I can rename it to deleteNamespaceResourcesAndRemoveNamespaceSpecFinalizer but thats a bit long ;)
I think the current name is fine (its a private func).
I dont want to change that code much.
There is a TODO to make it consistent with the kubernetes code. We should be able to reuse the same code in kubernetes and federation to delete namespace resources.

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Renamed to DeleteFromUnderlyingClusters

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Why?
My expectation is that if only the list of finalizers are being modified, then we should use Finalize(). Else use update().
Is that not correct?

File: federation/pkg/federation-controller/namespace/namespace_controller_test.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller_test.go]: Fixing in https://github.com/kubernetes/kubernetes/pull/35478

**Comment by nikhiljindal** [comment]
Verified that federation namespace e2e test is passing

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (caesarxuchao) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: No. `Finalize` is a special method that only exists in the namespace client. So I think we should call `Finalize()` only when we update the `namespace.spec.finalizer`.

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (caesarxuchao) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Ok. I still think "NamespaceSpecFinalizer" is more clear than "KubernetesFinalizer". Your call, as it's a private func.

> There is a TODO to make it consistent with the kubernetes code. We should be able to reuse the same code in kubernetes and federation to delete namespace resources.

Not sure I follow this part. What kubernetes code are you referring to?

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (caesarxuchao) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Ok. Thanks for the explanation.

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: oh ok. Hadnt realized the spec.Finalizer and ObjectMeta.Finalizer difference. You are right. Will update

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: To clarify, I named it removeKubernetesFinalizer because it removes the finalizer called FinalizerKubernetes.

> Not sure I follow this part. What kubernetes code are you referring to?

kubernetes namespace controller also has the same code that deletes all resources in a namespace before removing the finalizer - same as what this function does.

**Comment by nikhiljindal** [comment]
Replaced Finalize() by Update() as per the comment above and squashed commits.
PTAL.

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (nikhiljindal) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: Fixed

**Review by caesarxuchao — APPROVED** [review]

File: federation/pkg/federation-controller/namespace/namespace_controller.go
Comment (caesarxuchao) [inline federation/pkg/federation-controller/namespace/namespace_controller.go]: > To clarify, I named it removeKubernetesFinalizer because it removes the finalizer called FinalizerKubernetes.

I think `FinalizerKubernetes` is a bad name :)

**Comment by caesarxuchao** [comment]
LGTM. Thanks @nikhiljindal !

**Comment by nikhiljindal** [comment]
Trying to satisfy bazel and fix unit tests

**Comment by nikhiljindal** [comment]
All tests passed. Adding back LGTM

**Pull request merged**
